### ISSUE TITLE
docker daemon panic on push to registry

### ISSUE THREAD

**Issue by Zapadlo** [docker/distribution#1063]
Hello, attempting to set up a registry:2.0 with nginx auth in front of it. SSL terminates at the ELB. I can successfully log in and pull images, but on push docker daemon panics, stacktrace at the bottom.

`INFO[0000] Docker daemon  commit=d12ea79 execdriver=native-0.2 graphdriver=devicemapper version=1.8.1`
registry conf: https://github.com/Financial-Times/coco-registry-up
nginx proxy conf: https://github.com/Financial-Times/up-registry-proxy
```
daemon off;

worker_processes  1;

#error_log  logs/error.log;
#error_log  logs/error.log  notice;
#error_log  logs/error.log  info;

#pid        logs/nginx.pid;


events {
    worker_connections  1024;
}

http {
 # For versions of Nginx > 1.3.9 that include chunked transfer encoding support
 # Replace with appropriate values where necessary
 
 upstream docker-registry {
  server registry:5000;
 }
 
 server {
  listen 80 default_server;
 
  # ssl on;
  # ssl_certificate external/cert.pem;
  # ssl_certificate_key external/key.pem;
  
  # set HSTS-Header because we only allow https traffic
  # add_header Strict-Transport-Security "max-age=31536000; includeSubdomains";
  
  proxy_set_header Host       $http_host;   # required for Docker client sake
  # proxy_set_header X-Real-IP  $remote_addr; # pass on real client IP
 
  client_max_body_size 0; # disable any limits to avoid HTTP 413 for large image uploads
 
  # required to avoid HTTP 411: see Issue #1486 (https://github.com/dotcloud/docker/issues/1486)
  chunked_transfer_encoding on;
 
  location / {
      # let Nginx know about our auth file
      auth_basic              "Restricted";
      auth_basic_user_file    /docker-registry.htpasswd;
 
      proxy_pass http://docker-registry;
  }
  location /_ping {
      auth_basic off;
      proxy_pass http://docker-registry;
  }  
  location /v1/_ping {
      auth_basic off;
      proxy_pass http://docker-registry;
  }
  location /v2/ {
      if ($http_user_agent ~ "^(docker\/1\.(3|4|5(?!\.[0-9]-dev))|Go ).*\$" ) {
        return 404;
      }
 
      # To add basic authentication to v2 use auth_basic setting plus add_header
      auth_basic "Registry realm";
      auth_basic_user_file    /docker-registry.htpasswd;
      add_header 'Docker-Distribution-Api-Version' 'registry/2.0' always;
 
      proxy_pass                          http://docker-registry;
      proxy_set_header  Host              $http_host;   # required for docker client's sake
      # proxy_set_header  X-Real-IP         $remote_addr; # pass on real client's IP
      # proxy_set_header  X-Forwarded-For   $proxy_add_x_forwarded_for;
      # proxy_set_header  X-Forwarded-Proto $scheme;
      proxy_read_timeout                  900;
  }
 }
}
```

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x64381a]

goroutine 131 [running]:
bufio.(*Writer).flush(0xc20847b2c0, 0x0, 0x0)
	/usr/lib/go/src/bufio/bufio.go:530 +0xda
bufio.(*Writer).Flush(0xc20847b2c0, 0x0, 0x0)
	/usr/lib/go/src/bufio/bufio.go:519 +0x3a
net/http.(*response).Flush(0xc20818edc0)
	/usr/lib/go/src/net/http/server.go:1047 +0x4c
github.com/docker/docker/pkg/ioutils.(*WriteFlusher).Write(0xc2086167e0, 0xc208644840, 0xb6, 0x14a, 0xb6, 0x0, 0x0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/pkg/ioutils/writeflusher.go:21 +0x145
github.com/docker/docker/pkg/progressreader.updateProgress(0xc2085ab9d0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/pkg/progressreader/progressreader.go:63 +0x101
github.com/docker/docker/pkg/progressreader.(*Config).Read(0xc2085ab9d0, 0xc208ad4000, 0x8000, 0x8000, 0xf80380, 0x7f13c37dbe58, 0xc2087b36e0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/pkg/progressreader/progressreader.go:38 +0x229
io/ioutil.(*nopCloser).Read(0xc20878b700, 0xc208ad4000, 0x8000, 0x8000, 0xc20845f940, 0x0, 0x0)
	<autogenerated>:4 +0x89
io.(*multiReader).Read(0xc2087901c0, 0xc208ad4000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	/usr/lib/go/src/io/multi.go:13 +0xab
io.Copy(0x7f13c37f25a0, 0xc208779ae0, 0x7f13c37e6430, 0xc2087901c0, 0x48001, 0x0, 0x0)
	/usr/lib/go/src/io/io.go:362 +0x1f6
net/http.(*transferWriter).WriteBody(0xc20869c310, 0x7f13c37ed268, 0xc20845f940, 0x0, 0x0)
	/usr/lib/go/src/net/http/transfer.go:207 +0x226
net/http.(*Request).write(0xc2080c4b60, 0x7f13c37ed268, 0xc20845f940, 0xc20852f200, 0xc2087b30e0, 0x0, 0x0)
	/usr/lib/go/src/net/http/request.go:448 +0xada
net/http.(*persistConn).writeLoop(0xc2085268f0)
	/usr/lib/go/src/net/http/transport.go:951 +0x28d
created by net/http.(*Transport).dialConn
	/usr/lib/go/src/net/http/transport.go:661 +0xcbc

goroutine 1 [chan receive]:
main.(*DaemonCli).CmdDaemon(0xc20809b980, 0xc20800a020, 0x1, 0x1, 0x0, 0x0)
	/build/docker/src/docker/docker/daemon.go:292 +0x1811
reflect.callMethod(0xc20813e0f0, 0xc20812fce0)
	/usr/lib/go/src/reflect/value.go:605 +0x179
reflect.methodValueCall(0xc20800a020, 0x1, 0x1, 0x1, 0xc20813e0f0, 0x0, 0x0, 0xc20813e0f0, 0x0, 0x45102f, ...)
	/usr/lib/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc20813e030, 0xc20800a010, 0x2, 0x2, 0x0, 0x0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/cli/cli.go:89 +0x38e
main.main()
	/build/docker/src/docker/docker/docker.go:69 +0x428

goroutine 5 [syscall]:
os/signal.loop()
	/usr/lib/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/lib/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 10 [chan receive]:
github.com/docker/docker/api/server.(*Server).ServeApi(0xc20809ef40, 0xc20809bd80, 0x1, 0x1, 0x0, 0x0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/api/server/server.go:111 +0x74f
main.func·007()
	/build/docker/src/docker/docker/daemon.go:242 +0x5b
created by main.(*DaemonCli).CmdDaemon
	/build/docker/src/docker/docker/daemon.go:248 +0xd51

goroutine 11 [chan receive]:
github.com/docker/docker/daemon.func·025()
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/daemon/debugtrap_unix.go:17 +0x5c
created by github.com/docker/docker/daemon.setupDumpStackTrap
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/daemon/debugtrap_unix.go:20 +0x18e

goroutine 12 [IO wait]:
net.(*pollDesc).Wait(0xc2081ee060, 0x72, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2081ee060, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc2081ee000, 0x0, 0x7f13c37d9340, 0xc2081e04f0)
	/usr/lib/go/src/net/fd_unix.go:419 +0x40b
net.(*UnixListener).AcceptUnix(0xc2081f4120, 0xc2082202c0, 0x0, 0x0)
	/usr/lib/go/src/net/unixsock_posix.go:282 +0x56
net.(*UnixListener).Accept(0xc2081f4120, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go/src/net/unixsock_posix.go:293 +0x4c
github.com/docker/docker/pkg/listenbuffer.(*defaultListener).Accept(0xc2081f4140, 0x0, 0x0, 0x0, 0x0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/pkg/listenbuffer/buffer.go:71 +0x67
net/http.(*Server).Serve(0xc208040480, 0x7f13c37e1558, 0xc2081f4140, 0x0, 0x0)
	/usr/lib/go/src/net/http/server.go:1728 +0x92
github.com/docker/docker/api/server.(*HttpServer).Serve(0xc2081f44a0, 0x0, 0x0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/api/server/server.go:126 +0x4d
github.com/docker/docker/api/server.func·002(0x7f13c37e15b8, 0xc2081f44a0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/api/server/server.go:102 +0x40
created by github.com/docker/docker/api/server.(*Server).ServeApi
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/api/server/server.go:106 +0x6ac

goroutine 15 [IO wait]:
net.(*pollDesc).Wait(0xc2085aa300, 0x72, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2085aa300, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).readMsg(0xc2085aa2a0, 0xc2085bb220, 0x10, 0x10, 0xc2085c1220, 0x1000, 0x1000, 0xffffffffffffffff, 0x0, 0x0, ...)
	/usr/lib/go/src/net/fd_unix.go:296 +0x54e
net.(*UnixConn).ReadMsgUnix(0xc208538220, 0xc2085bb220, 0x10, 0x10, 0xc2085c1220, 0x1000, 0x1000, 0x51, 0xc2085bafd4, 0x4, ...)
	/usr/lib/go/src/net/unixsock_posix.go:147 +0x167
github.com/godbus/dbus.(*oobReader).Read(0xc2085c1200, 0xc2085bb220, 0x10, 0x10, 0xc2085c1200, 0x0, 0x0)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/transport_unix.go:21 +0xc5
io.ReadAtLeast(0x7f13c37e6408, 0xc2085c1200, 0xc2085bb220, 0x10, 0x10, 0x10, 0x0, 0x0, 0x0)
	/usr/lib/go/src/io/io.go:298 +0xf1
io.ReadFull(0x7f13c37e6408, 0xc2085c1200, 0xc2085bb220, 0x10, 0x10, 0x51, 0x0, 0x0)
	/usr/lib/go/src/io/io.go:316 +0x6d
github.com/godbus/dbus.(*unixTransport).ReadMessage(0xc2085934c0, 0xc2085a6ed0, 0x0, 0x0)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/transport_unix.go:85 +0x1bf
github.com/godbus/dbus.(*Conn).inWorker(0xc2085ad320)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/conn.go:248 +0x58
created by github.com/godbus/dbus.(*Conn).Auth
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/auth.go:118 +0xe84

goroutine 16 [chan receive]:
github.com/godbus/dbus.(*Conn).outWorker(0xc2085ad320)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/conn.go:370 +0x58
created by github.com/godbus/dbus.(*Conn).Auth
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/auth.go:119 +0xea1

goroutine 18 [chan receive]:
github.com/docker/libnetwork/iptables.signalHandler()
	/build/docker/src/docker/vendor/src/github.com/docker/libnetwork/iptables/firewalld.go:92 +0x57
created by github.com/docker/libnetwork/iptables.FirewalldInit
	/build/docker/src/docker/vendor/src/github.com/docker/libnetwork/iptables/firewalld.go:48 +0x185

goroutine 41 [chan receive]:
database/sql.(*DB).connectionOpener(0xc2085f3220)
	/usr/lib/go/src/database/sql/sql.go:589 +0x4c
created by database/sql.Open
	/usr/lib/go/src/database/sql/sql.go:452 +0x31c

goroutine 42 [IO wait]:
net.(*pollDesc).Wait(0xc20860cae0, 0x72, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc20860cae0, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).readMsg(0xc20860ca80, 0xc20863e480, 0x10, 0x10, 0xc208630820, 0x1000, 0x1000, 0xffffffffffffffff, 0x0, 0x0, ...)
	/usr/lib/go/src/net/fd_unix.go:296 +0x54e
net.(*UnixConn).ReadMsgUnix(0xc208538908, 0xc20863e480, 0x10, 0x10, 0xc208630820, 0x1000, 0x1000, 0x50, 0xc20863e234, 0x4, ...)
	/usr/lib/go/src/net/unixsock_posix.go:147 +0x167
github.com/godbus/dbus.(*oobReader).Read(0xc208630800, 0xc20863e480, 0x10, 0x10, 0xc208630800, 0x0, 0x0)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/transport_unix.go:21 +0xc5
io.ReadAtLeast(0x7f13c37e6408, 0xc208630800, 0xc20863e480, 0x10, 0x10, 0x10, 0x0, 0x0, 0x0)
	/usr/lib/go/src/io/io.go:298 +0xf1
io.ReadFull(0x7f13c37e6408, 0xc208630800, 0xc20863e480, 0x10, 0x10, 0x50, 0x0, 0x0)
	/usr/lib/go/src/io/io.go:316 +0x6d
github.com/godbus/dbus.(*unixTransport).ReadMessage(0xc208607550, 0xc2085fff80, 0x0, 0x0)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/transport_unix.go:85 +0x1bf
github.com/godbus/dbus.(*Conn).inWorker(0xc208614000)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/conn.go:248 +0x58
created by github.com/godbus/dbus.(*Conn).Auth
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/auth.go:118 +0xe84

goroutine 43 [chan receive]:
github.com/godbus/dbus.(*Conn).outWorker(0xc208614000)
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/conn.go:370 +0x58
created by github.com/godbus/dbus.(*Conn).Auth
	/build/docker/src/docker/vendor/src/github.com/godbus/dbus/auth.go:119 +0xea1

goroutine 44 [chan receive]:
github.com/coreos/go-systemd/dbus.func·001()
	/build/docker/src/docker/vendor/src/github.com/coreos/go-systemd/dbus/subscription.go:70 +0x64
created by github.com/coreos/go-systemd/dbus.(*Conn).initDispatch
	/build/docker/src/docker/vendor/src/github.com/coreos/go-systemd/dbus/subscription.go:94 +0x11c

goroutine 45 [chan receive]:
github.com/docker/docker/daemon.(*statsCollector).run(0xc208642390)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/daemon/stats_collector_unix.go:91 +0xb2
created by github.com/docker/docker/daemon.newStatsCollector
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/daemon/stats_collector_unix.go:31 +0x116

goroutine 46 [chan receive]:
github.com/docker/docker/daemon.(*Daemon).execCommandGC(0xc2080a20e0)
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/daemon/exec.go:256 +0x8c
created by github.com/docker/docker/daemon.NewDaemon
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/daemon/daemon.go:742 +0x2518

goroutine 130 [select]:
net/http.(*persistConn).readLoop(0xc2085268f0)
	/usr/lib/go/src/net/http/transport.go:928 +0x9ce
created by net/http.(*Transport).dialConn
	/usr/lib/go/src/net/http/transport.go:660 +0xc9f

goroutine 92 [chan receive]:
github.com/docker/docker/pkg/signal.func·002()
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/pkg/signal/trap.go:29 +0x8f
created by github.com/docker/docker/pkg/signal.Trap
	/build/docker/src/docker/.gopath/src/github.com/docker/docker/pkg/signal/trap.go:55 +0x250
```

**Comment by aaronlehmann** [comment]
This panic is a bug in the docker daemon. It was fixed in Docker 1.8.2 by https://github.com/docker/docker/pull/15353, though the code has changed a bit since then on the master branch.

The panic happens when a download fails, so it seems very possible that a configuration issue with the proxy is indirectly causing the panic.

I'd suggest upgrading to Docker 1.8.2 for further debugging.

**Comment by Zapadlo** [comment]
Thanks, we solved the issue by enabling debug logging on the nginx. It was a filesystem permissions problem. Closing for now, thanks!
